soci/store/store.go

```diff
@@ -0,0 +1,356 @@
+}
+
+type ContentStoreType string
+
+const (
+	ContainerdContentStoreType ContentStoreType = "containerd"
+	SociContentStoreType       ContentStoreType = "soci"
+	DefaultContentStoreType    ContentStoreType = SociContentStoreType
+)
+
+// ContentStoreTypes returns a slice of all supported content store types.
+func ContentStoreTypes() []ContentStoreType {
+	return []ContentStoreType{SociContentStoreType, ContainerdContentStoreType}
+}
+
+const (
+	// Default path to containerd content addressable storage
+	DefaultContainerdContentStorePath = "/var/lib/containerd/io.containerd.content.v1.content"
+
+	// Default path to soci content addressable storage
+	DefaultSociContentStorePath = "/var/lib/soci-snapshotter-grpc/content"
+)
+
+type Config struct {
+	Type      ContentStoreType `toml:"type"`
+	Namespace string           `toml:"namespace"`
+}
+
+func NewStoreConfig(opts ...Option) Config {
+	storeConfig := Config{
+		Type:      DefaultContentStoreType,
+		Namespace: namespaces.Default,
+	}
+	for _, o := range opts {
+		o(&storeConfig)
+	}
+	return storeConfig
+}
+
+type Option func(*Config)
+
+func WithNamespace(namespace string) Option {
+	return func(sc *Config) {
+		sc.Namespace = namespace
+	}
+}
+
+func WithType(contentStoreType ContentStoreType) Option {
+	return func(sc *Config) {
+		sc.Type = contentStoreType
+	}
+}
+
+func ErrUnknownContentStoreType(contentStoreType ContentStoreType) error {
+	return fmt.Errorf("unknown content store type: %s; must be one of %s or %s",
+		contentStoreType, ContainerdContentStoreType, SociContentStoreType)
+}
+
+// CanonicalizeContentStoreType resolves the empty string to DefaultContentStoreType,
+// returns other types, or errors on unrecognized types.
+func CanonicalizeContentStoreType(contentStoreType ContentStoreType) (ContentStoreType, error) {
+	switch contentStoreType {
+	case "":
+		return DefaultContentStoreType, nil
+	case ContainerdContentStoreType, SociContentStoreType:
+		return contentStoreType, nil
+	default:
+		return "", ErrUnknownContentStoreType(contentStoreType)
+	}
+}
+
+// GetContentStorePath returns the top level directory for the content store.
+func GetContentStorePath(contentStoreType ContentStoreType) (string, error) {
+	contentStoreType, err := CanonicalizeContentStoreType(contentStoreType)
+	if err != nil {
+		return "", err
+	}
+	switch contentStoreType {
+	case ContainerdContentStoreType:
+		return DefaultContainerdContentStorePath, nil
+	case SociContentStoreType:
+		return DefaultSociContentStorePath, nil
+	default:
+		return "", ErrUnknownContentStoreType(contentStoreType)
+	}
+}
+
+// GetContentStoreBlobPath returns the bottom level directory for the content store, e.g. "/blobs/sha256".
+func GetContentStoreBlobPath(contentStoreType ContentStoreType) (string, error) {
+	contentStorePath, err := GetContentStorePath(contentStoreType)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(contentStorePath, "blobs", "sha256"), nil
```

Comment:
nit: Since this is only used in tests maybe we can move it to util/testutl.

Reply:
Good call. This is very much an implementation detail of the content stores, and not something we want to export.